Adding options to specify class mappings

agrawalsourav98 opened the pull request:
Modified `convert_geojson.py` to take in class based on an attribute specified in the geojson without breaking backward compatibility.
The new features include:
1. Specify an attribute to be used as the class
2. Use default_class for single class conversions
3. Specify a mapping for the values, e.g. attribute has a value 1 which represents Apple class.
4. Specify a list of classes to be skipped

isaaccorley commented on ml_dronebase_data_utils/convert_geojson.py:
```diff
@@ -12,18 +12,35 @@
 
 
-def geo_to_voc(ortho_path: str, geo_path: str, save_path: str):
+def geo_to_voc(ortho_path: str, geo_path: str, save_path: str, class_attribute: str = None, class_mapping:dict = None,default_class:str = 'panel',skip_classes: list =[], rotated: bool = False):
```
Might be good to add a docstring to this function. There are many arguments now and it would help clarify what each arg does.

isaaccorley commented on test_conversion.py:
```diff
@@ -0,0 +1,26 @@
+from ml_dronebase_data_utils.convert_geojson import geo_to_voc
```
Need to use pytest for tests

jonlwowski012 commented on ml_dronebase_data_utils/convert_geojson.py:
```diff
@@ -12,46 +12,52 @@
+        # Dumb Logic
+        try:
+            # int(name) might be too restrictive in some scenarios, adapt if required
+            name = int(name)
```
What I normally do is make a list of all class names by appending class names to it if it doesnt exist in the list. Then I use the index as the class number. Does that make sense?

agrawalsourav98 replied:
So, The idea being I already have an id for each class and a name mapping for it. So, I think this makes more sense. Also, the int conversion is because the ids for some reason are float.

jonlwowski012 commented on ml_dronebase_data_utils/convert_geojson.py:
```diff
@@ -12,46 +12,52 @@
     if "s3://" in save_path:
         writer.save("annot.xml")
-        upload_file("annot.xml", save_path)
+        upload_file("annot.xml", save_path, exist_ok=False)
```
Should this always be false?

agrawalsourav98 replied:
When someone is rerunning it, I think the intended action he/she wants is to override the s3 location. We can add a flag if there might be other use cases.

jonlwowski012 commented on ml_dronebase_data_utils/convert_geojson.py:
```diff
@@ -60,8 +66,8 @@ def get_pixel_vertices(ortho: DatasetReader, gdf: GeoDataFrame) -> np.ndarray:
     pxl_vertices = []
-    for i, geo_v in enumerate(tqdm(geo_vertices, total=len(geo_vertices))):
-        pxl_v = [ortho.index(x, y) for x, y in geo_v]
+    for _, geo_v in enumerate(tqdm(geo_vertices, total=len(geo_vertices))):
```
Remove enumerate if we are not going to use index

jonlwowski012 commented on visualize_conversion.py:
```diff
@@ -0,0 +1,59 @@
+import tempfile
+
+if __name__ == "__main__":
+    ortho_path = "s3://ml-solar-ortho-fault-detection/orthos/tiff/PA140004_Thermal.tif"
```
Can we wrap this into a class and use argparse to instantiate ortho_path, annotation_path, and save_path?

jonlwowski012 commented:
Overall this PR looks good. Just some small things to fix before approving.

agrawalsourav98 commented:
Worked on the above comments and added the following features,
1. Added a CLI interface for convient way to process orthos and geojsons on s3 with support for batch processing

jonlwowski012 commented on ml_dronebase_data_utils/box_utils.py:
```diff
@@ -199,8 +199,21 @@ def boxes_to_vertices(boxes: Union[np.ndarray, List[List[float]]]) -> np.ndarray
     return np.asarray(vertices)
 
 
+"""
```
Comments typically go in the function

jonlwowski012 commented on ml_dronebase_data_utils/convert_geojson.py:
```diff
@@ -12,68 +13,103 @@
-        writer.save("annot.xml")
-        upload_file("annot.xml", save_path)
-        os.remove("annot.xml")
+        anno_path = os.path.basename(save_path)
```
Can you use os or Path to find the extension instead of assuming its the last 4 chars

agrawalsourav98 replied:
Yes

agrawalsourav98 replied:
Changes:
1. Merged both rotated and non-rotated geojson conversion into one function as most of the functionality was same apart from a few lines.
2. Added option to specify class mappings, which map id/numbers to string names
3. Added option to skip certain classes

agrawalsourav98 commented on ml_dronebase_data_utils/box_utils.py:
```diff
@@ -213,12 +226,20 @@ def rotated_boxes_to_vertices(
     sorted_idxs = np.argsort(-areas).tolist()
     # Re-order overlapped instances in descending order.
     boxes = boxes[sorted_idxs]
+    # Re-order classes as well
```
Sort classes here as boxes are also sorted here

agrawalsourav98 commented on ml_dronebase_data_utils/convert_geojson_cli.py:
```diff
@@ -0,0 +1,145 @@
+from pathlib import Path
```
Added CLI for geojson conversion. See README for usage.

agrawalsourav98 commented on ml_dronebase_data_utils/pascal_voc.py:
```diff
@@ -15,6 +15,7 @@ def __init__(
         depth: int = 3,
         database: str = "Unknown",
         segmented: int = 0,
+        prefix: str = "",
```
Adding a prefix which allows us to specify aws path even though conversion was done locally

agrawalsourav98 commented on ml_dronebase_data_utils/visualize.py:
```diff
@@ -9,32 +9,47 @@
 def draw_boxes(
     image: Union[Image.Image, np.ndarray],
     boxes: Optional[Union[np.ndarray, List[List[float]]]],
+    classes: List[str] = [],
```
Adding option to write class labels onto the vis images

agrawalsourav98 commented on ml_dronebase_data_utils/visualize_converted_geojson.py:
```diff
@@ -0,0 +1,164 @@
+import argparse
```
Added CLI for easy vis

agrawalsourav98 commented on setup.py:
```diff
@@ -45,10 +45,16 @@
             "flake8==4.0.1",
             "pytest",
         ],
+        entry_points={
```
Added entry points for the CLIs.

agrawalsourav98 commented on tests/test_conversion.py:
```diff
@@ -0,0 +1,38 @@
+from ml_dronebase_data_utils.convert_geojson import geo_to_voc
```
Added test for checking conversion

agrawalsourav98 commented on tests/test_s3.py:
```diff
@@ -31,9 +32,12 @@ def test_list_prefix_prefixes():
 def test_sync():
     data_url = (
-        "s3://ml-detectron-test-dataset/data/solar-panel-dataset/train_only_annots/"
+        "s3://ml-detectron-test-dataset/data/solar-panel-dataset/train/annotations/"
```
Fixed S3 location

agrawalsourav98 commented on tests/test_s3.py:
```diff
@@ -31,9 +32,12 @@ def test_list_prefix_prefixes():
     assert len(data_files) > 0
+
+    # Remove synced dir locally
+    shutil.rmtree("solar-panel-dataset-v2")
```
Added removal of downloaded folder after test

agrawalsourav98 commented on tests/test_visualization.py:
```diff
@@ -0,0 +1,9 @@
+from ml_dronebase_data_utils.visualize_converted_geojson import visualize
```
Added visualization test
